### PR TITLE
[WIP] Feature beat

### PR DESCRIPTION
**Pull request by chrismeyersfsu**
Beat replacement 

### ToDone:
* expire beat message so multiple tasks of the same type don't build up (amqp expiration)
* Account for beats that get skipped. This will happen if the time to service a beat (basically calling apply_async) takes long enough that it causes another tasks (or it's own) next beat to be missed. i.e. a super duper busy system.

### Won't do now:
* Set a timer for scheduled jobs

### CODE REVIEW

File: awx/settings/defaults.py [hunk 1]
```diff
@@ -477,10 +477,6 @@ def IS_TESTING(argv=None):
         'schedule': timedelta(seconds=60),
         'options': {'expires': 50,}
     },
-    'purge_stdout_files': {
```
Comment (ryanpetrello) [inline awx/settings/defaults.py]: We do still need this.

Comment (matburt) [reply awx/settings/defaults.py]: @rooftopcellist also is looking to put in a schedule for cleaning up expired tokens.

File: awx/main/management/commands/run_dispatcher.py [hunk 2]
```diff
@@ -35,48 +35,35 @@ def add_arguments(self, parser):
...
             def tick(self, *args, **kwargs):
                 if os.getppid() != self.ppid:
                     # if the parent PID changes, this process has been orphaned
                     # via e.g., segfault or sigkill, we should exit too
                     raise SystemExit()
                 return super(AWXScheduler, self).tick(*args, **kwargs)
 
-            def apply_async(self, entry, producer=None, advance=True, **kwargs):
+            def apply_async(self, entry, producer=None, advance=True, expiration=None, **kwargs):
                 for conn in connections.all():
```
Comment (ryanpetrello) [inline awx/main/management/commands/run_dispatcher.py:54]: Seems like this code could just go in the base implementation?  This function signature is unique to celery.

File: awx/main/beat/__init__.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        """
+        pass
+
+    def next_wakeup_seconds(self):
+        self.scheduler.update_time()
+        return (self.scheduler.next_wakeup() - self.scheduler.get_time()).total_seconds()
+
+    def run_tasks(self):
+        for sched in self.scheduler.get_schedules_to_run():
+            self.apply_async(sched, queue=get_local_queuename(),
```
Comment (ryanpetrello) [inline awx/main/beat/__init__.py:126]: This needs exception handling.  What happens if `apply_async` raises an uncaught exception?  If AMQP goes down?

**Comment by ryanpetrello** [comment]
My main reservations with merging this (now) is that beat _does_ work, and getting this wrong - even in a minor way - has the potential to cause _lots_ of weird, hard to diagnose problems.  I'd *really* want to battle test this change with full integration runs, I'd want to run VMs with this code for long stretches of time (days/weeks) to verify that jobs ran when they should, and I'd want very strenuous testing of edge cases, like:

1.  What happens if AMQP is randomly up/down?  Does this process recover and stay up?
2.  Same question, but with the database.
3.  What happens if the clock shifts/drifts _while_ the process is running (similar to what would happen during a DST switchover).
4. The beat implementation has persistence that backs its state up in a Python shelve (I assume so that it can remember the "last run" time in the event of process death?).  Is there a reason/use case for that which we're not considering here?

Have you made an attempt to just port over the bits of beat we care about?
https://github.com/celery/celery/blob/master/celery/beat.py
Assuming **that** is okay from a licensing perspective, it doesn't seem like _that much more_ code, and it's code that has served us well with predictable behavior for years.

All of this isn't to say that we can't _eventually_ move away from the celery implementation - I just want to make sure that when we do, we have the time and resources necessary to verify and feel very confident in a change like this.

**Comment by wwitzel3** [comment]
I agree here, I would much prefer to a see a port of beat.py/schedule.py right from Celery.

**Comment by AlanCoding** [comment]
> I would much prefer to a see a port of beat.py/schedule.py right from Celery.

_looks_

current uses look to be

https://github.com/ansible/awx/blob/3eaed52b83124bc37b3c762a77fedd018abbec2e/awx/main/management/commands/run_dispatcher.py#L38-L40

https://github.com/ansible/awx/blob/3007b9c66acca5c18dbd3aff522c5ba8d2569f8c/awx/settings/defaults.py#L8

we're on version 4.2.1, so this would be a port of code around:

https://github.com/celery/celery/blob/v4.2.1/celery/beat.py#L448

as well as

https://github.com/celery/celery/blob/v4.2.1/celery/schedules.py

and there will be some other messy parts of tearing that out of the rest of the code base, a lot is done around their construct of "apps".

**Comment by ryanpetrello** [comment]
Closing this due to age.
